### PR TITLE
New interface to rewrite table on certain schema change operations

### PR DESCRIPTION
**Pull request by zachmu**
Interface isn't quite final but this can be checked in. Complementary dolt changes are done and tested but want to get feedback on this approach first.

### CODE REVIEW

**Review by andrew-wm-arthur — COMMENTED** [review]
Seems reasonable. It seems at this stage that the engine is deciding when table rewrites happen, not the integrator?

File: sql/analyzer/batch.go [hunk 1]
```diff
@@ -119,9 +119,11 @@ func (b *Batch) evalOnce(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope
 		a.PushDebugContext(rule.Id.String())
 		next, same, err = rule.Apply(ctx, a, prev, scope, sel)
 		allSame = same && allSame
-		if next != nil && !same {
-			a.LogDiff(prev, next)
+		if next != nil {
 			a.LogNode(next)
+			// We should only do this if the result has changed, but some rules currently misbehave and falsely report nothing
```
Comment (max-hoffman) [inline sql/analyzer/batch.go:124]: abbreviating the output is so much more useful, even though catalog, subqueries, unions, ctes falsely report a change when there wasn't

File: sql/analyzer/validate_create_table.go [hunk 2]
```diff
@@ -54,10 +54,6 @@ func validateCreateTable(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope
 }
 
 func validateAlterColumn(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
```
Comment (max-hoffman) [inline sql/analyzer/validate_create_table.go:56]: maybe we should call this `resolveAlterColumn` now

File: sql/core.go [hunk 3]
```diff
@@ -323,9 +323,14 @@ type MultiDatabaser interface {
 	WithDatabaseProvider(DatabaseProvider) (Node, error)
 }
 
-// SchemaTarget is a node that has a target schema that can be set
+// SchemaTarget is a node that has a target schema that can be set during analysis. This is necessary because some
+// schema objects (things that involve expressions, column references, etc.) can only be reified during analysis. The
+// target schema is the schema of a table under a DDL operation, not the schema of rows returned by this node.
 type SchemaTarget interface {
```
Comment (max-hoffman) [inline sql/core.go:329]: `DDLSchemaTarget`?

File: sql/plan/alter_table.go [hunk 4]
```diff
@@ -418,6 +447,179 @@ func (i addColumnIter) Close(context *sql.Context) error {
...
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// addColumnToSchema returns a new schema and a set of projection expressions that when applied to rows from the old
+// schema will result in rows in the new schema.
+func addColumnToSchema(schema sql.Schema, column *sql.Column, order *sql.ColumnOrder) (sql.Schema, []sql.Expression, error) {
```
Comment (max-hoffman) [inline sql/plan/alter_table.go:513]: this would be great in the analyzer, FixFieldIndexesForExpressions screws that up?

**Review by max-hoffman — APPROVED** [review]
LGTM, couple nits, final interface condensing alters into a target schema/projections in a resolve rule is ideal

File: sql/plan/alter_table.go [hunk 5]
```diff
@@ -305,6 +320,9 @@ func (a AddColumn) WithExpressions(exprs ...sql.Expression) (sql.Node, error) {
 	a.targetSch = schemaWithDefaults(a.targetSch, exprs[:len(a.targetSch)])
 
 	unwrappedColDefVal, ok := exprs[len(exprs)-1].(*expression.Wrapper).Unwrap().(*sql.ColumnDefaultValue)
+
+	// *sql.Column is a reference type, make a copy before we modify it so we don't affect the original node
+	a.column = a.column.Copy()
```
[the remark this answers is not in the document]
Comment (zachmu) [reply sql/plan/alter_table.go:325]: No but column is defined as a pointer type which is harder to change

File: sql/analyzer/validate_create_table.go (hunk 2 above)
Comment (zachmu) [reply sql/analyzer/validate_create_table.go:56]: Good call

File: sql/plan/show_create_table.go [hunk 6]
```diff
@@ -37,6 +37,10 @@ type ShowCreateTable struct {
 	AsOf             sql.Expression
 }
 
+var _ sql.Node = (*ShowCreateTable)(nil)
+var _ sql.Expressioner = (*ShowCreateTable)(nil)
+var _ sql.SchemaTarget = (*ShowCreateTable)(nil)
```
[the remark this answers is not in the document]
Comment (zachmu) [reply sql/plan/show_create_table.go:42]: Because that's how we get fully resolved column defaults to them

File: sql/plan/alter_table.go (hunk 4 above)
Comment (zachmu) [reply sql/plan/alter_table.go:513]: It definitely belongs in the analyzer. My goal is to put everything there when I'm done with this round of changes.

File: sql/analyzer/batch.go (hunk 1 above)
Comment (zachmu) [reply sql/analyzer/batch.go:124]: We definitely do not want to abbreviate output if somebody sets the debug and verbose flags on the analyzer

File: sql/core.go (hunk 3 above)
Comment (zachmu) [reply sql/core.go:329]: I'll consider this in the next set of changes, still thinking about this name

File: sql/analyzer/batch.go (hunk 1 above)
Comment (max-hoffman) [reply sql/analyzer/batch.go:124]: "Abbreviate the output" = print the plan once not 300 times

**Pull request merged**